# Web Scraping Lab

You will find in this notebook some scrapy exercises to practise your scraping skills.

**Tips:**

- Check the response status code for each request to ensure you have obtained the intended contennt.
- Print the response text in each request to understand the kind of info you are getting and its format.
- Check for patterns in the response text to extract the data/info requested in each question.
- Visit each url and take a look at its source through Chrome DevTools. You'll need to identify the html tags, special class names etc. used for the html content you are expected to extract.

- [Selenium Doc](https://www.selenium.dev/documentation/)
- [Requests library](http://docs.python-requests.org/en/master/#the-user-guide) documentation 
- [Urllib](https://docs.python.org/3/library/urllib.html#module-urllib)
- [re lib](https://docs.python.org/3/library/re.html)
- [lxml lib](https://lxml.de/)
- [Scrapy](https://scrapy.org/)
- [List of HTTP status codes](https://en.wikipedia.org/wiki/List_of_HTTP_status_codes)
- [HTML basics](http://www.simplehtmlguide.com/cheatsheet.php)
- [CSS basics](https://www.cssbasics.com/#page_start)

#### Below are the libraries and modules you may need. `Selenium` and `pandas` are imported for you.

In [1]:
%pip install selenium

Note: you may need to restart the kernel to use updated packages.


In [2]:
%pip install webdriver-manager

Note: you may need to restart the kernel to use updated packages.


In [3]:
from selenium import webdriver

In [4]:
import pandas as pd
from selenium.webdriver.chrome.options import Options
from webdriver_manager.chrome import ChromeDriverManager

In [5]:
#driver configuration
opciones=Options()

opciones.add_experimental_option('excludeSwitches', ['enable-automation'])
opciones.add_experimental_option('useAutomationExtension', False)
opciones.headless=False    # si True, no aperece la ventana (headless=no visible)
opciones.add_argument('--start-maximized')         # comienza maximizado
#opciones.add_argument('user-data-dir=selenium')    # mantiene las cookies
#opciones.add_extension('driver_folder/adblock.crx')       # adblocker
opciones.add_argument('--incognito')

In [38]:
import warnings
warnings.filterwarnings('ignore')

In [39]:
PATH=ChromeDriverManager().install()





====== WebDriver manager ======
Current google-chrome version is 100.0.4896
Get LATEST chromedriver version for 100.0.4896 google-chrome
Driver [/Users/danigomezlechonbarrachina/.wdm/drivers/chromedriver/mac64/100.0.4896.60/chromedriver] found in cache


#### Download, parse, and print the content from the Trending Developers page from GitHub:

In [41]:
# This is the url you will scrape in this exercise
url = 'https://github.com/trending/developers'

In [42]:
driver=webdriver.Chrome(PATH, options = opciones)


driver.get(url)

In [43]:
tabla=driver.find_element_by_xpath('//*[@id="js-pjax-container"]/div[3]/div/div[2]')
filas=tabla.find_elements_by_tag_name('h1')


In [44]:
data=[e.text for e in  filas]

content=[]

for e in range(1,len(data),2):
    content.append(data[e])

In [45]:
content[:2]

['fullPage.js', 'golang-cheat-sheet']

#### Display the names of the trending developers retrieved in the previous step.

Your output should be a Python list of developer names. Each name should not contain any html tag.

**Instructions:**

1. Find out the html tag and class names used for the developer names.

1. Use string manipulation techniques to replace whitespaces and linebreaks (i.e. `\n`) in the *text* of each html element. Use a list to store the clean names.

1. Print the list of names.

Your output should look like below:

```
['trimstray (@trimstray)',
 'joewalnes (JoeWalnes)',
 'charlax (Charles-AxelDein)',
 'ForrestKnight (ForrestKnight)',
 'revery-ui (revery-ui)',
 'alibaba (Alibaba)',
 'Microsoft (Microsoft)',
 'github (GitHub)',
 'facebook (Facebook)',
 'boazsegev (Bo)',
 'google (Google)',
 'cloudfetch',
 'sindresorhus (SindreSorhus)',
 'tensorflow',
 'apache (TheApacheSoftwareFoundation)',
 'DevonCrawford (DevonCrawford)',
 'ARMmbed (ArmMbed)',
 'vuejs (vuejs)',
 'fastai (fast.ai)',
 'QiShaoXuan (Qi)',
 'joelparkerhenderson (JoelParkerHenderson)',
 'torvalds (LinusTorvalds)',
 'CyC2018',
 'komeiji-satori (神楽坂覚々)',
 'script-8']
 ```

In [49]:
data=[e.text for e in  filas]

names=[data[e] for e in range(0,len(data),2)]

In [50]:
names[:2]

['Álvaro', 'Ariel Mashraki']

In [51]:
driver.quit

<bound method ChromiumDriver.quit of <selenium.webdriver.chrome.webdriver.WebDriver (session="88536bc65e695442875e7af16036b22d")>>

#### Display the trending Python repositories in GitHub

The steps to solve this problem is similar to the previous one except that you need to find out the repository names instead of developer names.

In [52]:
# This is the url you will scrape in this exercise
url = 'https://github.com/trending/python?since=daily'

In [53]:
driver=webdriver.Chrome(PATH, options = opciones)
driver.get(url)

In [55]:
tabla=driver.find_element_by_xpath('//*[@id="js-pjax-container"]/div[3]/div/div[2]')
filas=tabla.find_elements_by_tag_name('a')


In [56]:
data=[e.text for e in  filas]

In [61]:
repos=[data[e] for e in range(len(data))  if "/" in data[e]]

In [63]:
repos[:2]

['fighting41love / funNLP', 'Sjj1024 / QiangCai']

In [65]:
driver.quit

<bound method ChromiumDriver.quit of <selenium.webdriver.chrome.webdriver.WebDriver (session="ffdc42b4b09ec1633c24c10d389e538b")>>

#### Display all the image links from Walt Disney wikipedia page

In [90]:
# This is the url you will scrape in this exercise
url = 'https://en.wikipedia.org/wiki/Walt_Disney'
driver=webdriver.Chrome(PATH, options = opciones)
driver.get(url)

In [132]:
from selenium.webdriver.common.by import By

In [133]:
img_tag = driver.find_elements(by = By.TAG_NAME, value = 'img')
imgs = [img.get_attribute('src') for img in img_tag]

In [135]:
imgs[:2]

['https://upload.wikimedia.org/wikipedia/en/thumb/e/e7/Cscr-featured.svg/20px-Cscr-featured.svg.png',
 'https://upload.wikimedia.org/wikipedia/en/thumb/8/8c/Extended-protection-shackle.svg/20px-Extended-protection-shackle.svg.png']

In [136]:
driver.quit

<bound method ChromiumDriver.quit of <selenium.webdriver.chrome.webdriver.WebDriver (session="250af8188380bdf18171892189a30990")>>

#### Retrieve an arbitary Wikipedia page of "Python" and create a list of links on that page

In [193]:
# This is the url you will scrape in this exercise
url ='https://en.wikipedia.org/wiki/Python' 
driver=webdriver.Chrome(PATH, options = opciones)
driver.get(url)

In [157]:
link_tag = driver.find_elements(by = By.TAG_NAME, value = 'a')
links = [l.get_attribute('href') for l in link_tag]

In [195]:
links[1:5]


['https://en.wikipedia.org/wiki/Python#mw-head',
 'https://en.wikipedia.org/wiki/Python#searchInput',
 'https://en.wiktionary.org/wiki/Python',
 'https://en.wiktionary.org/wiki/python']

In [163]:
driver.quit

<bound method ChromiumDriver.quit of <selenium.webdriver.chrome.webdriver.WebDriver (session="c5c87070756c3e8e5bbade8a0c25029b")>>

#### Number of Titles that have changed in the United States Code since its last release point 

In [203]:
# This is the url you will scrape in this exercise
url = 'http://uscode.house.gov/download/download.shtml'
driver=webdriver.Chrome(PATH, options = opciones)
driver.get(url)

In [221]:
#metodo 1
titles_tag = driver.find_elements(by = By.TAG_NAME, value = 'div.usctitlechanged')


In [223]:
data=[e.text for e in  titles_tag]

In [224]:
data

['Title 12 - Banks and Banking',
 'Title 19 - Customs Duties',
 'Title 22 - Foreign Relations and Intercourse']

#### A Python list with the top ten FBI's Most Wanted names 

In [225]:
# This is the url you will scrape in this exercise
url = 'https://www.fbi.gov/wanted/topten'
driver=webdriver.Chrome(PATH, options = opciones)
driver.get(url)

In [231]:
tabla=driver.find_element_by_xpath('//*[@id="main-content"]/div/div[2]/div')
filas=tabla.find_elements_by_tag_name('a')


In [233]:
data=[e.text for e in  filas]

In [242]:
names=[data[i] for i  in range(1,len(data),2)]

names[:3]

['RAFAEL CARO-QUINTERO', 'YULAN ADONAY ARCHAGA CARIAS', 'EUGENE PALMER']

####  20 latest earthquakes info (date, time, latitude, longitude and region name) by the EMSC as a pandas dataframe

In [243]:
# This is the url you will scrape in this exercise
url = 'https://www.emsc-csem.org/Earthquake/'
driver=webdriver.Chrome(PATH, options = opciones)
driver.get(url)

In [ ]:
#your code

#### Count number of tweets by a given Twitter account.

You will need to include a ***try/except block*** for account names not found. 
<br>***Hint:*** the program should count the number of tweets for any provided account

In [244]:
# This is the url you will scrape in this exercise 
# You will need to add the account credentials to this url
url = 'https://twitter.com/'
driver=webdriver.Chrome(PATH, options = opciones)
driver.get(url)

In [ ]:
#your code

#### Number of followers of a given twitter account

You will need to include a ***try/except block*** in case account/s name not found. 
<br>***Hint:*** the program should count the followers for any provided account

In [ ]:
# This is the url you will scrape in this exercise 
# You will need to add the account credentials to this url
url = 'https://twitter.com/'

In [ ]:
#your code

#### List all language names and number of related articles in the order they appear in wikipedia.org

In [245]:
# This is the url you will scrape in this exercise
url = 'https://www.wikipedia.org/'
driver=webdriver.Chrome(PATH, options = opciones)
driver.get(url)

In [246]:
titles_tag = driver.find_elements(by = By.TAG_NAME, value = 'strong')

In [252]:
data=[e.text for e in  titles_tag]
idiomas=[data[i] for i in range(1,len(data)-1)]
idiomas[:2]

['Español', 'English']

In [253]:
driver.quit

<bound method ChromiumDriver.quit of <selenium.webdriver.chrome.webdriver.WebDriver (session="d56db5fe8d8431f08675f749f11d6808")>>

#### A list with the different kind of datasets available in data.gov.uk 

In [254]:
# This is the url you will scrape in this exercise
url = 'https://data.gov.uk/'
driver=webdriver.Chrome(PATH, options = opciones)
driver.get(url)

In [257]:
sets_tag = driver.find_elements(by = By.TAG_NAME, value = 'a.govuk-link')


In [263]:
sets=[e.text for e in  sets_tag]
sets=sets[4:]
sets[:4]

['Business and economy', 'Crime and justice', 'Defence', 'Education']

In [264]:
driver.quit

<bound method ChromiumDriver.quit of <selenium.webdriver.chrome.webdriver.WebDriver (session="45d9c6b96eabba143d60dde85fecf320")>>

#### Top 10 languages by number of native speakers stored in a Pandas Dataframe

In [265]:
# This is the url you will scrape in this exercise
url = 'https://en.wikipedia.org/wiki/List_of_languages_by_number_of_native_speakers'
driver=webdriver.Chrome(PATH, options = opciones)
driver.get(url)

In [266]:
tabla=driver.find_element_by_xpath('//*[@id="mw-content-text"]/div[1]/table[4]')
filas=tabla.find_elements_by_tag_name('a')



In [273]:
langs=[e.text for e in  filas]
langs[:4]

['Mandarin', 'Spanish', 'English', 'Hindi']

In [274]:
driver.quit

<bound method ChromiumDriver.quit of <selenium.webdriver.chrome.webdriver.WebDriver (session="2fb277545806cee5684ea872a5d56e1c")>>

### BONUS QUESTIONS

#### Scrape a certain number of tweets of a given Twitter account.

In [ ]:
# This is the url you will scrape in this exercise 
# You will need to add the account credentials to this url
url = 'https://twitter.com/'

In [ ]:
# your code

#### IMDB's Top 250 data (movie name, Initial release, director name and stars) as a pandas dataframe

In [ ]:
# This is the url you will scrape in this exercise 
url = 'https://www.imdb.com/chart/top'

In [ ]:
# your code

#### Movie name, year and a brief summary of the top 10 random movies (IMDB) as a pandas dataframe.

In [ ]:
#This is the url you will scrape in this exercise
url = 'http://www.imdb.com/chart/top'

In [ ]:
#your code

#### Find the live weather report (temperature, wind speed, description and weather) of a given city.

In [ ]:
#https://openweathermap.org/current
city = city=input('Enter the city:')
url = 'http://api.openweathermap.org/data/2.5/weather?'+'q='+city+'&APPID=b35975e18dc93725acb092f7272cc6b8&units=metric'

In [ ]:
# your code

#### Book name,price and stock availability as a pandas dataframe.

In [ ]:
# This is the url you will scrape in this exercise. 
# It is a fictional bookstore created to be scraped. 
url = 'http://books.toscrape.com/'

In [ ]:
#your code